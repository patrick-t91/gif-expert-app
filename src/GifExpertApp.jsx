import { useState } from "react";
import {  GifGrid, AddCategory } from "./components";

const GifExpertApp = () => {
  const [categories, setCategories] = useState(["One Punch"]);

  const onAddCategory = (newCategory) => {
    // Tarea: agregar una nueva categoria
    // setCategories([...categories, !true ? "Nothing" : "Slum Dunk"]); // Podemos aplicar logica para agregar elementos
    // Otra forma de actualizar estado:
    // setCategories(cat => [...cat, 'Slam Dunk'])

    /*Antes de React v18, si habia un setState dentro de una funcion, se renderizaba el 
    componente una vez por cada setState y ademas por la funcion. Ahora solo se renderiza 
    al final de la funcion, lo cual hace a React aun mas rapido de lo que ya era */
    if (categories.includes(newCategory)) return;
    setCategories([...categories, newCategory]);
  };

  return (
    <>
      <h1>GifExpertApp</h1>

      <AddCategory
        // setCategories={setCategories}
        // onNewCategory es una forma de emitir un valor del componente hijo (AddCategory) al padre (GifExpertApp)
        onNewCategory={onAddCategory} // onNewCategory={(value)=>onAddCategory(value)} al haber un solo parametro, no hace falta pasarlo
      />
      <button onClick={() => onAddCategory()}>Agregar</button>
      {categories.map((category) => (
        <GifGrid category={category} key={category} />
      ))}
    </>
  );
};

export default GifExpertApp;
