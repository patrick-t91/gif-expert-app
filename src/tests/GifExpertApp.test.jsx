import { render, screen, fireEvent } from "@testing-library/react";
import GifExpertApp from "../GifExpertApp";

describe("Pruebas en GifExpertApp", () => {
  test("Debe coincidir con el snapshot", () => {
    const { container } = render(<GifExpertApp />);
    expect(container).toMatchSnapshot();
  });

  test("Debe cambiar el valor del input", () => {
    render(<GifExpertApp />);

    const input = screen.getByRole("textbox");
    fireEvent.input(input, { target: { value: "Saitama" } });

    expect(input.value).toBe("Saitama");
  });

  test("Debe mostrar la categoria agregada", () => {
    const inputValue = "One Punch";

    render(<GifExpertApp />);

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");
    fireEvent.input(input, { target: { value: inputValue } });
    fireEvent.submit(form);

    expect(screen.getByText(inputValue));
  });

  test("No deberia agregarse una categoria mas de una vez", () => {
    const inputValue = "One Punch";

    render(<GifExpertApp />);

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");
    fireEvent.input(input, { target: { value: inputValue } });
    fireEvent.submit(form);
    fireEvent.input(input, { target: { value: inputValue } });
    fireEvent.submit(form);

    expect(screen.getAllByText(inputValue).length).toBe(1);
  });
});
