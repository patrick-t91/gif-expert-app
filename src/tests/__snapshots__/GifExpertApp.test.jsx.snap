// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifExpertApp Debe coincidir con el snapshot 1`] = `
<div>
  <h1>
    GifExpertApp
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar Gifs"
      type="text"
      value="One Punch"
    />
  </form>
  <button>
    Agregar
  </button>
  <h3>
    One Punch
  </h3>
  <h3>
    Cargando...
  </h3>
  <div
    class="card-grid"
  />
</div>
`;
